### PR TITLE
Detect BuildTools to compile with visual studio

### LINKED ISSUE #2256: CMake Error: Generator: execution of make failed. Make command was: "devenv.com" when only VS 2017 Build tools are installed
Hi,

first off, I'm a conan fan. Great job, you guys!

I'm using conan 0.30.3 on Windows 10 and I am using Visual Studio 2017 (up to date, as of Jan 4).

I have an issue when trying to create a package from the eigen project. Please find all files here:
https://github.com/philsuess/testProject

On my PC, where I have the full Visual Studio 2017 (Professional), everything runs fine with these commands:

`set CONAN_USERNAME=optconan`
`conan create %CONAN_USERNAME%/testing -s compiler="Visual Studio" -s compiler.version=15 -s build_type=Release`

However, on my build machine (a VM running Windows 10, with conan 0.30.3, but only Visual Studio Build Tools (15.5.2) installed), I get several errors trying to call 'devenv.cmd' like the following:

`CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_daa4f"`

Now, as I understand, `devenv` is not available if only the Build Tools are installed. At least I can not find it anywhere in my installation. In `build_sln_command` has the option `upgrade_project` default to true, so that `devenv` is part of the command. Should this not be the case if I only have Build Tools installed? -I'm assuming that this is the cause for the error.

Can you help shed some light?

Here is the full output from the erroneous build on the VM:

```
Running with gitlab-runner 10.2.0 (0a75cdd1)
  on cirunnerwin.conan (cc76d47c)
Using Shell executor...
Running on CIRUNNERWIN...
Fetching changes...
HEAD is now at a6e0786 remove license check
From https://gitlab.****/conan-Eigen
 * [new tag]         v3.3.4     -> v3.3.4
Checking out a4744d3e as master...
Skipping Git submodules setup
$ set CONAN_USERNAME=optconan
$ set CONAN_CHANNEL=
$ conan create %CONAN_USERNAME%/testing -s compiler="Visual Studio" -s compiler.version=15 -s build_type=Release -s compiler.runtime="MD"
eigen/3.3.4@optconan/testing: Exporting package recipe
eigen/3.3.4@optconan/testing export: Copied 1 '.txt' files: CMakeLists.txt
eigen/3.3.4@optconan/testing: The stored package has not changed
eigen/3.3.4@optconan/testing: Installing package
Requirements
    eigen/3.3.4@optconan/testing from local
Packages
    eigen/3.3.4@optconan/testing:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9

eigen/3.3.4@optconan/testing: WARN: Forced build from source
eigen/3.3.4@optconan/testing: Building your package in E:\conan\.conan\data\eigen\3.3.4\optconan\testing\build\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
eigen/3.3.4@optconan/testing: Copying sources to build folder
eigen/3.3.4@optconan/testing: Generator cmake created conanbuildinfo.cmake
eigen/3.3.4@optconan/testing: Calling build()
-- The C compiler identification is MSVC 19.12.25830.2
-- The CXX compiler identification is MSVC 19.12.25830.2
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x64/cl.exe
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x64/cl.exe -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x64/cl.exe
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x64/cl.exe -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: called by CMake conan helper
-- Conan: Using cmake global configuration
-- Conan: Adjusting default RPATHs Conan policies
-- Performing Test standard_math_library_linked_to_automatically
-- Performing Test standard_math_library_linked_to_automatically - Success
-- Standard libraries to link to explicitly: none
-- Performing Test COMPILER_SUPPORT_OPENMP
-- Performing Test COMPILER_SUPPORT_OPENMP - Success
-- Performing Test EIGEN_COMPILER_SUPPORT_CPP11
-- Performing Test EIGEN_COMPILER_SUPPORT_CPP11 - Success
-- Found unsuitable Qt version "" from NOTFOUND
-- Performing Test EIGEN_COMPILER_SUPPORT_CXX11
-- Performing Test EIGEN_COMPILER_SUPPORT_CXX11 - Success
-- The Fortran compiler identification is unknown
-- A cache variable, namely BLAS_DIR, has been set to specify the install directory of BLAS
-- Looking for BLAS -- mkl.h not found
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE  
-- Looking for MKL BLAS: not found
-- Looking for Goto BLAS: not found
-- Looking for Open BLAS: not found
-- Looking for Eigen BLAS: not found
-- Looking for Eigen BLAS: not found
-- Looking for Atlas BLAS: not found
-- Looking for PhiPACK BLAS: not found
-- Looking for CXML BLAS: not found
-- Looking for DXML BLAS: not found
-- Looking for SunPerf BLAS: not found
-- Looking for SCSL BLAS: not found
-- Looking for SGIMATH BLAS: not found
-- Looking for IBM ESSL BLAS: not found
-- Looking for IBM ESSL MT BLAS: not found
-- Looking for ACML BLAS: not found
-- Looking for ACML BLAS: not found
-- Looking for ACML BLAS: not found
-- Looking for Apple BLAS: not found
-- Looking for NAS BLAS: not found
-- Looking for Generic BLAS: not found
-- Looking for Generic BLAS: not found
-- Looking for Generic BLAS: not found
-- Looking for Generic BLAS: not found
-- Looking for Generic BLAS: not found
CMake Warning at eigen/cmake/FindBLAS.cmake:1371 (message):
  BLA_VENDOR has been set to All but blas libraries could not be found or
  check of symbols failed.

  Please indicate where to find blas libraries.  You have three options:

  - Option 1: Provide the installation directory of BLAS library with cmake
  option: -DBLAS_DIR=your/path/to/blas

  - Option 2: Provide the directory where to find BLAS libraries with cmake
  option: -DBLAS_LIBDIR=your/path/to/blas/libs

  - Option 3: Update your environment variable (Linux: LD_LIBRARY_PATH,
  Windows: LIB, Mac: DYLD_LIBRARY_PATH)

  

  To follow libraries detection more precisely you can activate a verbose
  mode with -DBLAS_VERBOSE=ON at cmake configure.

  You could also specify a BLAS vendor to look for by setting
  -DBLA_VENDOR=blas_vendor_name.

  List of possible BLAS vendor: Goto, ATLAS PhiPACK, CXML, DXML, SunPerf,
  SCSL, SGIMATH, IBMESSL, Intel10_32 (intel mkl v10 32 bit),Intel10_64lp
  (intel mkl v10 64 bit, lp thread model, lp64 model), Intel10_64lp_seq
  (intel mkl v10 64 bit, sequential code, lp64 model),Intel( older versions
  of mkl 32 and 64 bit), ACML, ACML_MP, ACML_GPU, Apple, NAS, Generic
Call Stack (most recent call first):
  eigen/cmake/FindLAPACK.cmake:138 (find_package)
  eigen/test/CMakeLists.txt:30 (find_package)


-- A library with BLAS API not found. Please specify library location.
-- LAPACK requires BLAS.
-- Could NOT find CHOLMOD (missing: CHOLMOD_INCLUDES CHOLMOD_LIBRARIES) 
-- Could NOT find UMFPACK (missing: UMFPACK_INCLUDES UMFPACK_LIBRARIES) 
-- Could NOT find SUPERLU (missing: SUPERLU_INCLUDES SUPERLU_LIBRARIES SUPERLU_VERSION_OK) 
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
-- Could NOT find BLAS (missing: BLAS_SEQ_LIBRARIES BLAS_LIBRARY_DIRS) 
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_daa4f"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_9fc51"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_807f7"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_481ab"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_880b4"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_40940"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_b015a"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_e3c60"
-- Performing Test SCOTCH_Num_4
-- Performing Test SCOTCH_Num_4 - Failed
-- Performing Test SCOTCH_Num_8
-- Performing Test SCOTCH_Num_8 - Failed
-- Could NOT find MPI_C (missing: MPI_C_WORKS) 
-- Could NOT find MPI_CXX (missing: MPI_CXX_COMPILER) 
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_23edc"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_fea82"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_dd9bb"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_dafa0"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_f771c"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_eec12"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_ab174"
CMake Error: Generator: execution of make failed. Make command was: "devenv.com" "CMAKE_TRY_COMPILE.sln" "/build" "Debug" "/project" "cmTC_dbbd1"
-- Could NOT find MPI_Fortran (missing: MPI_Fortran_COMPILER) 
-- Could NOT find MPI (missing: MPI_C_FOUND MPI_CXX_FOUND MPI_Fortran_FOUND) 
-- Performing Test PTSCOTCH_Num_4
-- Performing Test PTSCOTCH_Num_4 - Failed
-- Performing Test PTSCOTCH_Num_8
-- Performing Test PTSCOTCH_Num_8 - Failed
-- A version of Pastix has been found but pastix_nompi.h does not exist in the include directory. Because Eigen tests require a version without MPI, we disable the Pastix backend.
-- Could NOT find SPQR (missing: SPQR_INCLUDES SPQR_LIBRARIES) 
-- Found unsuitable Qt version "" from NOTFOUND
-- Performing Test COMPILER_SUPPORT_FASTMATH
-- Performing Test COMPILER_SUPPORT_FASTMATH - Failed
-- Performing Test COMPILER_SUPPORT_FPFAST
-- Performing Test COMPILER_SUPPORT_FPFAST - Success
-- Could NOT find Boost
-- Could NOT find GOOGLEHASH (missing: GOOGLEHASH_INCLUDES GOOGLEHASH_COMPILE) 
-- Could NOT find ADOLC (missing: ADOLC_INCLUDES ADOLC_LIBRARIES) 
-- Could NOT find MPFR (missing: MPFR_INCLUDES MPFR_LIBRARIES MPFR_VERSION_OK) (Required is at least version "2.3.0")
-- Could NOT find GMP (missing: GMP_INCLUDES GMP_LIBRARIES) 
-- Could NOT find PkgConfig (missing: PKG_CONFIG_EXECUTABLE) 
-- Could NOT find FFTW (missing: FFTW_INCLUDES FFTW_LIBRARIES) 
-- Found OpenGL: opengl32   
-- Could NOT find GLUT (missing: GLUT_glut_LIBRARY GLUT_INCLUDE_DIR) 
CUDA_TOOLKIT_ROOT_DIR not found or specified
-- Could NOT find CUDA (missing: CUDA_TOOLKIT_ROOT_DIR CUDA_NVCC_EXECUTABLE CUDA_INCLUDE_DIRS CUDA_CUDART_LIBRARY) (Required is at least version "7.0")
-- Found unsuitable Qt version "" from NOTFOUND
-- Qt4 not found, so disabling the mandelbrot and opengl demos
-- ************************************************************
-- ***    Eigen's unit tests configuration summary          ***
-- ************************************************************
-- 
-- Build type:        Release
-- Build site:        cirunnerwin
-- Build string:      unknownwin-19.12.25830.2-sse2-64bit
-- Enabled backends:  
-- Disabled backends: Cholmod,  UmfPack,  SuperLU,  PaStiX,  METIS,  SPQR,  Qt4 support,  Boost.Multiprecision,  GoogleHash,  Adolc,  MPFR C++,  fftw,  OpenGL, 
-- Default order:     Column-major
-- Maximal matrix/vector size: 320
-- SSE2:              Using architecture defaults
-- SSE3:              Using architecture defaults
-- SSSE3:             Using architecture defaults
-- SSE4.1:            Using architecture defaults
-- SSE4.2:            Using architecture defaults
-- AVX:               Using architecture defaults
-- FMA:               Using architecture defaults
-- AVX512:            Using architecture defaults
-- Altivec:           Using architecture defaults
-- VSX:               Using architecture defaults
-- ARM NEON:          Using architecture defaults
-- ARMv8 NEON:        Using architecture defaults
-- S390X ZVECTOR:     Using architecture defaults
-- C++11:             OFF
-- SYCL:              OFF
-- CUDA:              OFF
-- 
CXX:               C:/Program Files (x86)/Microsoft Visual Studio/2017/BuildTools/VC/Tools/MSVC/14.12.25827/bin/Hostx86/x64/cl.exe
 CXX_FLAGS:         /DWIN32 /D_WINDOWS /W4 /GR /EHsc  /MP2 /EHsc /wd4127 /wd4505 /wd4714 /D_CRT_SECURE_NO_WARNINGS /D_SCL_SECURE_NO_WARNINGS
 Sparse lib flags:   

-- ************************************************************
-- 
-- Configured Eigen 3.3.4
-- 
-- To build/run the unit tests, read this page:
--   http://eigen.tuxfamily.org/index.php?title=Tests
-- 
-- Configuring incomplete, errors occurred!
See also "E:/conan/.conan/data/eigen/3.3.4/optconan/testing/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/CMakeFiles/CMakeOutput.log".
See also "E:/conan/.conan/data/eigen/3.3.4/optconan/testing/build/5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9/CMakeFiles/CMakeError.log".
eigen/3.3.4@optconan/testing: 
eigen/3.3.4@optconan/testing: ERROR: Package '5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9' build failed
eigen/3.3.4@optconan/testing: WARN: Build folder E:\conan\.conan\data\eigen\3.3.4\optconan\testing\build\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9
ERROR: eigen/3.3.4@optconan/testing: Error in build() method, line 39
	cmake.configure(build_dir=".")
	ConanException: Error 1 while executing cd . && cmake -G "Visual Studio 15 2017 Win64" -DCONAN_LINK_RUNTIME="/MD" -DCONAN_EXPORTED="1" -DCONAN_COMPILER="Visual Studio" -DCONAN_COMPILER_VERSION="15" -DCMAKE_INSTALL_PREFIX="E:\conan\.conan\data\eigen\3.3.4\optconan\testing\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9" -DCONAN_CXX_FLAGS="/MP2" -DCONAN_C_FLAGS="/MP2" -Wno-dev E:\conan\.conan\data\eigen\3.3.4\optconan\testing\source
ERROR: Job failed: exit status 1`

```

### LINKED ISSUE #2258: Only Windows BuildTools installed
Here in package tools there is an issue related with an installation of only the buildtools, but not the whole visual studio, investigate how the vcvars command is working (or failing) and how vswhere.exe could help:

https://github.com/conan-io/conan-package-tools/pull/137

### PR DESCRIPTION
**Pull request by danimtb**
closes #2256, closes #2258 (conan does not detect BuildTools).

Related to #2279, #1683

- Added new helper function to use `vswhere` in tools
- Improved tools.vs_installation_path(): Now it can also locate BuildTools installations.

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://raw.githubusercontent.com/conan-io/conan/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. Also adding a description of the changes in the ``changelog.rst`` file. https://github.com/conan-io/docs

I will open a PR to docs documenting the changes in both tools after this is reviewed.

### CODE REVIEW

File: conans/client/tools/win.py [hunk 1]
```diff
@@ -67,25 +68,91 @@ def vs_installation_path(version):
...
+        for installation in output:
+            if installation["installationVersion"].startswith(("%d." % int(version))):
+                vs_path = installation["installationPath"]
 
         # Remember to cache result
         vs_installation_path._cached[version] = vs_path
 
     return vs_installation_path._cached[version]
 
+def vswhere(all_=False, prerelease_=False, products_=list(), requires_=list(), version_="",
```
Comment (memsharded) [inline conans/client/tools/win.py]: Do not use a list object as default parameter, it is a bug. Use None, then inside the method initialize it

Comment (memsharded) [reply conans/client/tools/win.py]: Why all those underscores for argument names?

File: conans/client/tools/win.py [hunk 2]
```diff
@@ -67,25 +68,91 @@ def vs_installation_path(version):
...
+    if all_:
+        arguments.append("-all")
+
+    if prerelease_:
+        arguments.append("-prerelease")
+
+    if products_:
+        arguments.append("-products")
+
+        for product in products_:
```
Comment (memsharded) [inline conans/client/tools/win.py]: ``arguments.extend(products_)``

File: conans/client/tools/win.py (hunk 1 above)
Comment (danimtb) [reply conans/client/tools/win.py]: Search for "single_trailing_underscore_" in [PEP 8](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles)

Comment (memsharded) [reply conans/client/tools/win.py]: But that is only ``all``, the rest are not python words, are they?

Comment (danimtb) [reply conans/client/tools/win.py]: all, format and property

**Comment by danimtb** [comment]
Pending: add try/catch to `tools.vswhere()`

File: conans/client/tools/win.py [hunk 3]
```diff
@@ -68,26 +69,94 @@ def vs_installation_path(version):
...
-            if os.path.isfile(vswhere_path):
-                version_range = "[%d.0, %d.0)" % (int(version), int(version) + 1)
-                try:
-                    output = subprocess.check_output([vswhere_path, "-version", version_range,
-                                                      "-legacy", "-property", "installationPath"])
-                    vs_path = output.decode().strip()
-                    _global_output.info("vswhere detected VS %s in %s" % (version, vs_path))
-                except (ValueError, subprocess.CalledProcessError, UnicodeDecodeError) as e:
-                    _global_output.error("vswhere error: %s" % str(e))
+        output_legacy = json.loads(vswhere(legacy=True, format_="json"))
```
Comment (lasote) [inline conans/client/tools/win.py]: vswhere should return the data already loaded, so move the `json.loads` to tools.vswhere

File: conans/client/tools/win.py [hunk 4]
```diff
@@ -68,26 +69,94 @@ def vs_installation_path(version):
...
         # Remember to cache result
         vs_installation_path._cached[version] = vs_path
 
     return vs_installation_path._cached[version]
 
 
+def vswhere(all_=False, prerelease=False, products=None, requires=None, version="", latest=False,
+            legacy=False, format_="", property_="", nologo=True):
+
+    # 'version' option only works if Visual Studio 2017 is installed:
```
Comment (lasote) [inline conans/client/tools/win.py]: What is the output of this function? list of paths?

Comment (danimtb) [reply conans/client/tools/win.py]: I could be a list of `products` (json objects with lots of fileds: `installationPath`, `productId`...), or if there is just one product it returns just that json object.

Comment (lasote) [reply conans/client/tools/win.py]: Then I think we should unify the return and return always a list, even with one element, check the return type is a nightmare

File: conans/client/tools/win.py [hunk 5]
```diff
@@ -68,26 +69,92 @@ def vs_installation_path(version):
...
-            if os.path.isfile(vswhere_path):
-                version_range = "[%d.0, %d.0)" % (int(version), int(version) + 1)
-                try:
-                    output = subprocess.check_output([vswhere_path, "-version", version_range,
-                                                      "-legacy", "-property", "installationPath"])
-                    vs_path = output.decode().strip()
-                    _global_output.info("vswhere detected VS %s in %s" % (version, vs_path))
-                except (ValueError, subprocess.CalledProcessError, UnicodeDecodeError) as e:
-                    _global_output.error("vswhere error: %s" % str(e))
+        output_legacy = vswhere(legacy=True)
```
Comment (lasote) [inline conans/client/tools/win.py]: maybe the "output" name in the vars is not very good. legacy_products and all_products?

File: conans/client/tools/win.py [hunk 6]
```diff
@@ -68,26 +69,92 @@ def vs_installation_path(version):
...
+        output_products = vswhere(products=["*"])
+        output = output_legacy + output_products
+
+        for installation in output:
```
Comment (lasote) [inline conans/client/tools/win.py]: `for product in products` could be nicer

**Pull request merged**